rviz_common/include/rviz_common/message_filter_display.hpp

```diff
@@ -103,6 +103,27 @@ class MessageFilterDisplay : public _RosTopicDisplay
 protected:
   void updateTopic() override
   {
+    // Change the QoS according to the info in the ROS Graph
+    const auto & publishers = rviz_ros_node_.lock()->get_raw_node()->get_publishers_info_by_topic(
+      topic_property_->getTopicStd());
+
+    if (!publishers.empty()) {
+      // Check if there is any Subscriber with the same QoS as set
+      bool any_publisher_qos_compatible = false;
+      for (const auto & publisher : publishers) {
+        if (qos_profile == publisher.qos_profile()) {
+          any_publisher_qos_compatible = true;
+          break;
+        }
+      }
+
+      // If not found, select one valid
+      if (!any_publisher_qos_compatible) {
+        qos_profile = publishers[0].qos_profile();
+        qos_profile_property_->setQoSProfile(qos_profile);
+      }
```

Comment:
This is not great, because e.g. if you pick `RELIABLE` reliability and there was another published that used `BEST_EFFORT`, you're going to only match the first one.
Whereas if you picked `BEST_EFFORT` for the subscription, that was going to match both publishers.

In practice, having more than one publisher in the same topic with different reliability is not a good idea, and when that's found the best thing to do is warn the user.
Anyways if we need to pick one default in those cases, it should be the one that matches the most cases.
For durability, if we find publishers with different settings we should pick VOLATILE, which matches all publisher.

Reply:
I think you can find an example of this logic in https://github.com/ros2/ros2cli/pull/653